File: frontend/src/types/research_intelligence.ts
export interface PublicationTrendPoint {
  year: number;
  publication_count: number;
  growth_rate: number | null;
}

export interface PublicationTrendsResponse {
  total_publications: number;
  year_range: {
    min_year: number | null;
    max_year: number | null;
  };
  points: PublicationTrendPoint[];
}

export interface DomainYearPoint {
  year: number;
  count: number;
  share_percentage: number;
}

export interface DomainTrendItem {
  domain: string;
  total_publications: number;
  average_citations: number;
  yearly_distribution: DomainYearPoint[];
}

export interface DomainTrendsResponse {
  total_domains: number;
  domains: DomainTrendItem[];
}

export interface KeywordYearPoint {
  year: number;
  count: number;
}

export interface KeywordTrendItem {
  keyword: string;
  total_occurrences: number;
  recent_count: number;
  yearly_distribution: KeywordYearPoint[];
}

export interface KeywordTrendsResponse {
  total_keywords: number;
  keywords: KeywordTrendItem[];
}

export interface TopCitedPublicationSummary {
  id: number;
  title: string;
  citation_count: number;
  publication_date: string | null;
  venue: string | null;
  primary_domain: string | null;
  doi: string | null;
}

export interface CitationYearPoint {
  year: number;
  total_citations: number;
  average_citations: number;
  publication_count: number;
}

export interface CitationStatisticsResponse {
  total_publications: number;
  total_citations: number;
  average_citations: number;
  median_citations: number;
  max_citations: number;
  citations_by_year: CitationYearPoint[];
  top_cited_publications: TopCitedPublicationSummary[];
}

export interface EmergingTopicItem {
  topic: string;
  recent_count: number;
  historical_count: number;
  growth_rate: number | null;
  velocity_score: number;
  status: 'EMERGING' | 'ESTABLISHED_GROWING' | 'STABLE';
  reasons: string[];
}

export interface EmergingTopicsResponse {
  evaluation_window: {
    recent_start_year: number;
    current_year: number;
    recent_window_length_years: number;
  };
  total_emerging: number;
  topics: EmergingTopicItem[];
}

export interface ResearchHotspotItem {
  domain_or_topic: string;
  category: 'domain' | 'keyword';
  publication_count: number;
  recent_growth_rate: number | null;
  citation_count: number;
  average_citations: number;
  hotspot_score: number;
  classification: 'CRITICAL_HOTSPOT' | 'HIGH_ACTIVITY' | 'MODERATE_ACTIVITY' | 'EMERGING_NICHE';
  key_drivers: string[];
}

export interface ResearchHotspotsResponse {
  total_hotspots: number;
  evaluation_timestamp: string;
  hotspots: ResearchHotspotItem[];
}

File: backend/app/schemas/research_intelligence.py
from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field


# --- Publication Trends ---
class PublicationTrendPoint(BaseModel):
    year: int
    publication_count: int
    growth_rate: Optional[float] = Field(None, description="Year-over-Year growth percentage")


class PublicationTrendsResponse(BaseModel):
    total_publications: int
    year_range: Dict[str, Optional[int]]
    points: List[PublicationTrendPoint]


# --- Domain Trends ---
class DomainYearPoint(BaseModel):
    year: int
    count: int
    share_percentage: float


class DomainTrendItem(BaseModel):
    domain: str
    total_publications: int
    average_citations: float
    yearly_distribution: List[DomainYearPoint]


class DomainTrendsResponse(BaseModel):
    total_domains: int
    domains: List[DomainTrendItem]


# --- Keyword Trends ---
class KeywordYearPoint(BaseModel):
    year: int
    count: int


class KeywordTrendItem(BaseModel):
    keyword: str
    total_occurrences: int
    recent_count: int
    yearly_distribution: List[KeywordYearPoint]


class KeywordTrendsResponse(BaseModel):
    total_keywords: int
    keywords: List[KeywordTrendItem]


# --- Citation Analytics ---
class TopCitedPublicationSummary(BaseModel):
    id: int
    title: str
    citation_count: int
    publication_date: Optional[datetime] = None
    venue: Optional[str] = None
    primary_domain: Optional[str] = None
    doi: Optional[str] = None


class CitationYearPoint(BaseModel):
    year: int
    total_citations: int
    average_citations: float
    publication_count: int


class CitationStatisticsResponse(BaseModel):
    total_publications: int
    total_citations: int
    average_citations: float
    median_citations: float
    max_citations: int
    citations_by_year: List[CitationYearPoint]
    top_cited_publications: List[TopCitedPublicationSummary]


# --- Emerging Topics ---
class EmergingTopicItem(BaseModel):
    topic: str
    recent_count: int
    historical_count: int
    growth_rate: Optional[float]
    velocity_score: float
    status: str = Field(..., description="EMERGING | ESTABLISHED_GROWING | STABLE")
    reasons: List[str] = []


class EmergingTopicsResponse(BaseModel):
    evaluation_window: Dict[str, Any]
    total_emerging: int
    topics: List[EmergingTopicItem]


# --- Research Hotspots ---
class ResearchHotspotItem(BaseModel):
    domain_or_topic: str
    category: str = Field(..., description="domain | keyword")
    publication_count: int
    recent_growth_rate: Optional[float]
    citation_count: int
    average_citations: float
    hotspot_score: float = Field(..., ge=0.0, le=100.0)
    classification: str = Field(..., description="CRITICAL_HOTSPOT | HIGH_ACTIVITY | MODERATE_ACTIVITY | EMERGING_NICHE")
    key_drivers: List[str] = []


class ResearchHotspotsResponse(BaseModel):
    total_hotspots: int
    evaluation_timestamp: datetime
    hotspots: List[ResearchHotspotItem]
